### PR TITLE
fix: information_schema.columns.datetime_precision

### PR DESCRIPTION
**Pull request by cyjake**
`new Model(values)` should have values rounded according to precision

### CODE REVIEW

File: src/raw.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+'use strict';
+
+module.exports = class Raw {
+  constructor(value) {
+    this.value = value;
+    // consumed in expr_formatter.js
+    this.type = 'raw';
+  }
+};
```
Comment (cyjake) [inline src/raw.js:9]: bit worried that the original raw data structure might be supplied by user input directly, such as `where({ name: { __raw: true, value: "'' OR 1" })`. Hence changed it to a specific js class that needs instantiation.

File: test/integration/suite/basics.test.js [hunk 2]
```diff
@@ -172,20 +172,20 @@ describe('=> Basic', () => {
     });
 
     it('Bone.previousChanged(): raw VS rawPrevious', async function () {
-      const post = new Post({ title: 'Unknown', extra: 'hello' });
+      const post = new Post({ title: 'Unknown', extra: { greeting: 'hello' } });
```
Comment (cyjake) [inline test/integration/suite/basics.test.js:175]: 'hello' won't be able to be converted into js object

File: src/drivers/postgres/schema.js [hunk 3]
```diff
@@ -67,6 +69,7 @@ module.exports = {
         // https://www.postgresql.org/docs/9.5/infoschema-table-constraints.html
         primaryKey,
         unique: row.constraint_type === 'UNIQUE',
+        datetimePrecision: precision === 6 ? null : precision,
```
Comment (cyjake) [inline src/drivers/postgres/schema.js:72]: postgres has datetime_precision default to 6 even if the precision is not specified in column definition. reset it back to `null` to work around existing test cases for now.

File: src/data_types.js [hunk 4]
```diff
@@ -374,7 +385,12 @@ class JSON extends DataType {
     if (!value) return value;
     // type === JSONB
     if (typeof value === 'object') return value;
-    return global.JSON.parse(value);
+    try {
+      return global.JSON.parse(value);
+    } catch (err) {
+      console.error(new Error(`unable to cast ${value} to JSON`));
+      return value;
+    }
```
Comment (cyjake) [inline src/data_types.js:393]: JSON.cast will now fallback to original value

File: src/data_types.js [hunk 5]
```diff
@@ -229,14 +229,24 @@ class DATE extends DataType {
...
-    if (precision > 0) return `${dataType}(${precision})`;
+    if (precision != null && precision >= 0) return `${dataType}(${precision})`;
     return dataType;
   }
 
+  _round(value) {
+    const { precision } = this;
+    if (precision != null && precision < 3 && value instanceof Date) {
+      const result = new Date(value);
+      result.setMilliseconds(result.getMilliseconds() % (10 ** precision));
```
Comment (JimmyDaddy) [inline src/data_types.js:240]: `result.getMilliseconds() % (10 ** precision)` seems not correct?
while `precision  = 2, Milliseconds = 864`, expect `86`
```js
864 % (10 ** 2) = 64
```
how about `Number(result.getMilliseconds()/ (10 ** (3- precision)).toFixed())`

Comment (cyjake) [reply src/data_types.js]: nice catch! updated with following solution:

```js
const divider = 10 ** (3 - precision);
 return new Date(Math.round(value.getTime() / divider) * divider);
```

**Review by JimmyDaddy — APPROVED** [review]

**Pull request merged**